packages/devtools_app/lib/src/debugger/span_parser.dart

```diff
@@ -0,0 +1,548 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:collection';
+import 'dart:convert';
+
+import 'package:string_scanner/string_scanner.dart';
+
+abstract class SpanParser {
+  /// Takes a TextMate [Grammar] and a [String] and outputs a list of
+  /// [ScopeSpan]s corresponding to the parsed input.
+  static List<ScopeSpan> parse(Grammar grammar, String src) {
+    final scanner = LineScanner(src);
+    final spans = <ScopeSpan>[];
+    while (!scanner.isDone) {
+      bool foundMatch = false;
+      for (final pattern in grammar.topLevelMatchers) {
+        final result = pattern.scan(grammar, scanner);
+        if (result != null) {
+          spans.addAll(result);
+          foundMatch = true;
+          break;
+        }
+      }
+      if (!foundMatch && !scanner.isDone) {
+        // Found no match, move forward by a character and try again.
+        scanner.readChar();
+      }
+    }
+    return spans;
+  }
+}
+
+/// A representation of a TextMate grammar used to create [ScopeSpan]s
+/// representing scopes within a body of text.
+/// 
+/// References used:
+///   - Grammar specification: 
+///       https://macromates.com/manual/en/language_grammars#language_grammars
+///   - Helpful blog post which clears up ambiguities in the spec: 
+///       https://www.apeth.com/nonblog/stories/textmatebundle.html
+/// 
+class Grammar {
+  factory Grammar.fromJson(Map<String, dynamic> json) {
+    return Grammar._(
+      name: json['name'],
+      scopeName: json['scopeName'],
+      topLevelMatchers:
+          _parseMatchers(json['patterns'].cast<Map<String, dynamic>>()),
+      repository: Repository.build(json),
+    );
+  }
+
+  Grammar._({
+    this.name,
+    this.scopeName,
+    this.topLevelMatchers,
+    this.repository,
+  });
+
+  static List<_Matcher> _parseMatchers(List<Map<String, dynamic>> patterns) {
+    final List<_Matcher> parsed = [];
+
+    void _parseMatchersHelper(List<_Matcher> patternList) {
+      if (patternList == null) {
+        return;
+      }
+      for (final pattern in patternList) {
+        if (pattern is _MultilineMatcher) {
+          _parseMatchersHelper(pattern.patterns);
```

Comment:
should this be `_parseMatchers`?

Reply:
Actually, it looks like this is just old code. I'll remove it.